81 add osm builder

garyservin opened the pull request:
@basicNew I've added the initial implementation for the OSM builder, can you please review it?

./terminus/run_generator.py --builder=OsmCityBuilder --parameters osm_map=/home/gary/terminus/terminus/samples/osm_sample/osrf.osm

I wasn't able to generate the svg (the visualizer crashes with segmentation fault), but the gazebo can load the generated sdf without issues.

basicNew commented on terminus/builders/osm_city_builder.py:
```diff
@@ -0,0 +1,86 @@
+        city = City()
+
+        # Ground plane
+        city.set_ground_plane(GroundPlane(100, Point(0, 0, 0)))
```
Shouldn't we make the ground plane the same size as the bounding box?

basicNew commented on terminus/builders/osm_city_builder.py:
```diff
@@ -0,0 +1,86 @@
+
+# For now exclude smaller roads, in the future we should draw each
+# type of street with different width/texture
+road_types = ['motorway', 'trunk', 'primary', 'secondary', 'tertiary',
```
Not much of a fan of having constants like this, esp. for testing. I know is a less pythonic way, but can we move this to a method in `OsmCityBuilder`?

basicNew commented on terminus/builders/osm_city_builder.py:
```diff
@@ -0,0 +1,86 @@
+            for ref in value['refs']:
+                x, y = self._translate_coords(self.osm_coords[ref]['lat'],
+                                              self.osm_coords[ref]['lon'])
+                tmp_road.add_node(JunctionNode.on(x, y, 0))
```
We should only create a junction when we know there are at least two streets crossing. If this is just a waypoint in the middle of the road but there is no intersection a `SimpleNode` should be used (also note that creating `JunctionNode`s only make sense if you are going to share the same instance between two or more roads; otherwise the RNDF generator will not be able to connect the entry/exit waypoints).

garyservin commented:
@basicNew I've addressed your comments and pushed the changes, I've also included a check so coords/nodes outside of the bounding box won't be used.

basicNew commented on terminus/builders/osm_city_builder.py:
```diff
@@ -0,0 +1,122 @@
+            tmp_road = Street()
+            for ref in value['refs']:
+                # Check if coord is inside bounding box
+                if self.bounds['minlat'] < self.osm_coords[ref]['lat'] < self.bounds['maxlat'] and \
```
hmm, I think we may have an issue here if the road starts inside the bounding box, goes out and then in again (think e.g. of a roundabout). I think that if we only trim the edges of the street we should be ok, but we have to avoid removing intermediate points, otherwise we will get a fake geometry.

garyservin replied:
yes, that makes sense, I'm not sure though if the reference points that osm uses are in "order" meaning that the first is the beginning of the street and the last is the end of the street. I'll investigate a bit

basicNew replied:
Ok, but consider that they should be in order in our representation, otherwise models won't work (and the visualization would be very weird)

basicNew commented:
Great, thanks @garyservin ! Since we are at a quick-prototyping stage we can skip tests now, but we should create at least an issue to add unit tests as soon as we have a version with junctions working.

garyservin commented:
@basicNew I've added some logic to make sure that roads that snake in and out of the bounding box get properly used. It is a very "hacky" solution, please let me know if you have a better idea on how to implement it. I've also added a mapfile that shows an example of this (attached an example image, look at the "Starr Way").

![2016-12-06-150136_3286x1080_scrot](https://cloud.githubusercontent.com/assets/475165/20937275/03830916-bbc5-11e6-9ba6-590e867a846d.png)

![sample_in_and_out](https://cloud.githubusercontent.com/assets/475165/20937379/7581e942-bbc5-11e6-9ccb-b92692f250d2.png)

basicNew commented:
This is great, thanks @garyservin ! Regarding the implementation, I actually thought it just as trimming a string:

- Create the road with all the points.
- Starting from the front of `road.nodes`, remove all points that lay outside the box. Stop as soon as you find the first one inside.
- Starting from the back of `road.nodes`, remove all points that lay outside the box. Stop as soon as you find the first one inside.

If it is ok with you, lets create a Technical debt issue to improve the readability of that part and we can merge this one now.

garyservin commented:
yes, that makes sense
